### ISSUE TITLE
Buffer not fully flushed when metric throughput is higher than the metric_batch_size / flush_interval and has large spikes 

### ISSUE THREAD

**Issue by gnjack** [influxdata/telegraf#9726]
I've explained the issue in more detail in the PR that seems to have caused it https://github.com/influxdata/telegraf/pull/8953#issuecomment-913711114

### Relevant telegraf.conf:
```toml
[agent]
  flush_interval = "10s"
  flush_jitter = "5s"
  interval = "1s"
  metric_batch_size = 2500
  metric_buffer_limit = 50000
  round_interval = true
  
[[outputs.influxdb_v2]]
  ...
```

### System info:

Behaviour seen after upgrading from Telegraf 1.16.3 to 1.19.1

### Steps to reproduce:

Have an average metric throughput greater than `metric_batch_size` / `flush_interval`. For example with a batch size of 2500 and a flush_interval of 10s, then write more than 250 metrics per second. This will trigger early flushes.

Have large spikes of metrics all gathered at once, that are several times greater than `metric_batch_size`, for example 10,000 metrics at the same time with a `metric_batch_size` of 2,500.

### Expected behavior:

When an early flush is triggered, all possible batches in the buffer are flushed, not just a single batch.
OR
Even when early flushes are triggered, the regular `flush_interval` flush is still triggered flushing all possible batches in the buffer.

### Actual behavior:

Large spikes in metrics cause only a **single** early flush of a **single** batch despite a buffer containing many batches of metrics. These buffered batches can build up resulting in a metric buffer overflow unless the normal `flush_interval` flush is called instead of an early flush.

You can see this on the following graphs - metrics build up in the buffer unflushed after every spike in metrics gathered, until we get lucky and `flush_interval` passes without an early flush. A normal `flush_interval` flush actually flushes all buffered batches in one method call, unlike early flushes which only flush one batch.
![image](https://user-images.githubusercontent.com/5956751/132237580-1be8543e-1bbd-45d1-9fee-879b10e94cc9.png)

### Additional info:

Appears to be caused by changes in this PR https://github.com/influxdata/telegraf/pull/8953#issuecomment-913711114 which was merged into telegraf 1.19.0

**Comment by powersj** [comment]
@gnjack if possible, could you attempt to use the Telegraf artifacts in my [PR](https://github.com/influxdata/telegraf/pull/9800#issuecomment-925298140)?

**Comment by Hipska** [comment]
Solved by #9800
